core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java

```diff
@@ -1814,6 +1814,14 @@ default OptionalInt getMaxWriterTasks(ConnectorSession session)
         return OptionalInt.empty();
     }
 
+    /**
+     * @return whether connector supports reading a single column from the give tableHandle, without reading the entire table.
+     */
+    default boolean isColumnarTableScan(ConnectorSession session, ConnectorTableHandle tableHandle)
```

Comment:
`ConnectorTableHandle` can represent a table or a relation (see JDBC pushdowns).  Also, selected columns are conveyed to the connector via applyProjection. Is it possible to rephrase the javadoc with these in mind?

Also, given applyProjection was already invoked, why does the engine care about other columns that are not going to be read? Maybe knowing this would help formulate the javadoc best.

Reply:
This method should return true only for columnar data formats that allow efficient reads of single columns i.e. reading column `a` separately from column `b` requires more or less the same effort as reading columns `a` and `b` together.
For this to be useful, projection pushdown also needs to be supported by the connector.
@findepi Maybe you can suggest a better javadoc with this in mind?
@Dith3r there is a typo there _give_ -> _given_

Reply:
SPI does not (and should not) know about file formats.
What is the SPI way to read `a` column separate from `b` , or read them together?
My guess is that it's very implicit and related to lazy blocks being returned from the page source. Is this correct?

Reply:
> My guess is that it's very implicit and related to lazy blocks being returned from the page source. Is this correct?

There are multiple levels of "columnar" reads:
1. Physically reading columns separately (that usually not happens as reads are merged)
2. Decoding columns separately.

This SPI is telling that cost of reading `a, b` is more or less sum of cost of reading `a` and `b` separately (when pruning projections are applied), so maybe that's better Javadoc for it

Reply:
We don't have a SPI method that takes a table handle (supposed to return `a,b`) and returns two separate table handles, one returning `a` and the other `b` column.

Reply:
> We don't have a SPI method that takes a table handle (supposed to return a,b) and returns two separate table handles, one returning a and the other b column.

This is what this PR is effectively doing

Reply:
what about

```
@return true if reading subset of columns from given table separately from reading a compliment of the subset has similar or better
performance as reading this table.
```

?

Reply:
Looks good to me with a small correction. You probably meant _complement_?

Reply:
This is a poor proxy for indicating whether a read is cheaper. The fact that we have to explain it in the javadocs means the method semantics are not very well specified -- and the explanation has nothing to do with the name of the method.

If the goal is to establish the cost of reading a set of columns, we should model it using cost estimations like we do elsewhere. I haven't thought enough about it, but that's the direction I would take. Maybe do a comparison between a given table handle that encompasses columns `a` and `b` (from the examples above) and the sum of the costs of pushing `a` and `b` separately into the table scan.

Reply:
The goal is not to establish a generic cost of reading but rather to determine if splitting a read into two reads does not add much overhead. That is what basically only columnar storage provides, hence the name.
Having the cost would of course work here but it is too complex for the use case and would make the implementation harder than necessary. For example, we don't need to compare the cost of reading two different columns here.

Reply:
> but rather to determine if splitting a read into two reads does not add much overhead. 

That's too specific for the SPI. And still, whether the read is columnar is not necessarily indicative of that. There are many other factors involved, including whether processing metadata is expensive, how many rows are read, how many columns there are in the table vs how many are being read, I/O costs if the columns are scattered, how much data is being read from underlying storage vs decompressed vs decoded, etc.

Reply:
Maybe you could add `Estimate dataSize` along with `io.trino.spi.statistics.TableStatistics#rowCount`. Could be unknown all connectors than "columnar" connectors. For "columnar" connectors it could be sum of dataSize of all involved columns. 
This way you could compare data size involved in table scan of `AB` with scan of `A` and `B` separately. 
wdyt?

Reply:
We may not know the data sizes until eg split generation, or even later.
I am OK with "determine if splitting a read into two reads does not add much overhead" concept.
I see how being more precise could be helpful, but it could force implementors into faking answers.

Reply:
> We may not know the data sizes until eg split generation, or even later.

Sure, it would be an estimate same as for column data size

Reply:
In Iceberg, estimate for a column size requires walking the manifest files and then processing data file information from them. Depending on number of files in a table, this is not cheap and definitely not free. Meanwhile, Iceberg can blindly answer the "whether splitting a read into two reads does not add much overhead" question.

Reply:
Do we have any conclusion which path is right one?

Reply:
One aspect we didn't discuss here is that with JDBC connectors there may be an expensive to compute join/aggregation pushed into the table handle and we don't want to repeat that computation only to project different column out of it for each sub-query.
That scenario is more problematic compared to wasted read from csv/text/avro etc.
How about we name this a bit differently
```
public boolean allowSplittingReadIntoMultipleSubQueries(ConnectorSession session, ConnectorTableHandle tableHandle)
``` 
A bit clunky but makes it clearer that it's more about making a judgement whether we want to split a read from table handle rather than about columnar/row orientation of data.

Reply:
makes sense to me. wdyt @martint ?